lib/likelihood.c

```diff
@@ -0,0 +1,304 @@
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * Implementation of log-likelihood functions for full-ARG.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+#include "likelihood.h"
+#include "msprime.h"
+
+int
+msp_increment_edge(tsk_treeseq_t *ts, tsk_edge_t *edge, tsk_node_t *parent, 
+                   tsk_node_t *child) {
+    int ret = 1;
+    ret = tsk_treeseq_get_edge(ts, edge->id + 1, edge);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->child, child);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->parent, parent);
+    if (ret != 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+int
+msp_unnormalised_log_likelihood_mut(tsk_treeseq_t *ts, double theta, 
+                                    double *lik) {
+    int ret = 1;
+    int it;
+    unsigned int j;
+    tsk_mutation_t mut;
+    tsk_size_t num_mutations = tsk_treeseq_get_num_mutations(ts);
+    tsk_id_t node_id;
+    double total_material = 0;
+    double branch_length;
+    tsk_tree_t *tree = NULL;
+    tsk_edge_t *edge = NULL;
+    tsk_node_t *parent = NULL;
+    tsk_node_t *child = NULL;
+    tree = malloc(sizeof *tree);
+    if (tree == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    edge = malloc(sizeof *edge);
+    if (edge == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    parent = malloc(sizeof *parent);
+    if (parent == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    child = malloc(sizeof *child);
+    if (child == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = tsk_treeseq_get_edge(ts, 0, edge);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->child, child);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->parent, parent);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_tree_init(tree, ts, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    do {
+        total_material += (edge->right - edge->left) 
+            * (parent->time - child->time);
+    } while (msp_increment_edge(ts, edge, parent, child) == 0);
+    if (theta > 0) {
+        *lik = num_mutations * log(total_material * theta) -
+            total_material * theta;
+        for (it = tsk_tree_first(tree); it == 1; it = tsk_tree_next(tree)) {
+            for (j = 0; j < tree->sites_length; j++) {
+                mut = tree->sites[j].mutations[0];
+                ret = tsk_treeseq_get_node(ts, mut.node, child);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_tree_get_parent(tree, child->id, &node_id);
```

Comment:
Might as well use ``tree->parent[child]`` here, we're sure the IDs are good so we don't nee the error checking.